### PR TITLE
Allow client certificate

### PR DESCRIPTION
**Pull request by jkamon**
Allow setting client private key to enable mTLS. Basically this is extension of #3958

### CODE REVIEW

**Comment by anuraaga** [comment]
We should add these options to the OpenTelemetry specification. Can you open a PR to add two variables here?

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#configuration-options

One for client private key, one for client certificate.

For a programmatic version we would just expect two variables, `setClientTls(byte[] privateKeyPem, byte[] certificatePem)`, the second could be a certificate chain in one payload. In the future we would probably add keystore support as well but can start with this after making the spec change.

**Comment by jkamon** [comment]
PR for spec is waiting for approval again. 
I also made changes to this PR to match names.

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/TlsUtil.java
Comment (jack-berg) [inline exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/TlsUtil.java:138]:
```
try (BufferedReader bufferedReader = Files.newBufferedReader(new File(filePath).toPath(), UTF_8)) {
  ...
}
```

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/TlsUtil.java
Comment (jack-berg) [inline exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/TlsUtil.java:146]: What happens if the file is malformed and ends before we find `-----END`? Looks like a NPE would get thrown once we reach the end of the file.

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/grpc/ManagedChannelUtil.java
Comment (jack-berg) [inline exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/grpc/ManagedChannelUtil.java:56]: Looks like there's a formatting issue, i.e. it should be `privatekKeyChainPem != null`.

You can run `./gradlew spotlessApply` to apply the formatting conventions.

File: exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/TlsUtilsTest.java
Comment (jack-berg) [inline exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/TlsUtilsTest.java:10]: #nit: Can be package private. Also the class is named `TlsUtil`, rather than `TlsUtils`

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java:174]: `readFileBytes(...)`?

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java:181]: How about exiting early if `filePath == null` here?
```
if (filePath == null) {
 return null;
}
...
```

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java:188]: Can just return here and save a few lines:
```
return Files.readAllBytes(path)
```

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/GrpcServiceBuilder.java
Comment (jack-berg) [inline sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/GrpcServiceBuilder.java:29]: Have people asked for client cert support for jaeger remote sampling?

File: exporters/jaeger/src/test/java/io/opentelemetry/exporter/jaeger/JaegerGrpcSpanExporterTest.java
Comment (jkamon) [reply exporters/jaeger/src/test/java/io/opentelemetry/exporter/jaeger/JaegerGrpcSpanExporterTest.java]: removed

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/TlsUtil.java
Comment (jkamon) [reply exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/TlsUtil.java]: added

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/TlsUtil.java
Comment (jkamon) [reply exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/TlsUtil.java]: added

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/TlsUtil.java
Comment (jkamon) [reply exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/TlsUtil.java]: added check also for this case

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/grpc/ManagedChannelUtil.java
Comment (jkamon) [reply exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/grpc/ManagedChannelUtil.java]: done

File: exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/TlsUtilsTest.java
Comment (jkamon) [reply exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/TlsUtilsTest.java]: modifier changed. renamed.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java
Comment (jkamon) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java]: not sure if understood. suggesting method rename? done.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java
Comment (jkamon) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java:182]: Existing code tries to load data according to specific property (log, metric, span) and if not existing then generic one without datatype is used. Exiting early would change this logic and not use generic property. 
Maybe I didn't understood this proposal but for now I will keep this part as it is. Could you please elaborate a little more?

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java
Comment (jkamon) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java]: fixed

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/GrpcServiceBuilder.java
Comment (jkamon) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/GrpcServiceBuilder.java]: I have tried to add client cert everywhere where server certificate could be used. 
Is there any problem with having this?

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java
Comment (jack-berg) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java:182]: If you exit early with, the the code inside this if block doesn't need to be indented:
```
if (filePath == null) {
  return null;
}
Path path = Paths.get(filePath);
if (!Files.exists(path)) {
  throw new ConfigurationException(
      "Invalid file: " + path + " (configured in property: " + propertyToRead + ")");
}
try {
  return Files.readAllBytes(path);
} catch (IOException e) {
  throw new ConfigurationException(
      "Error reading content of file (" + path + ") configured in " + propertyToRead, e);
}
```

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java:111]: I believe you're missing a condition. If `clientKeyBytes == null && clientKeyChainBytes == null` we shouldn't call `setClientTls.accept(..)`.

File: exporters/jaeger/src/test/java/io/opentelemetry/exporter/jaeger/JaegerGrpcSpanExporterTest.java
Comment (jack-berg) [inline exporters/jaeger/src/test/java/io/opentelemetry/exporter/jaeger/JaegerGrpcSpanExporterTest.java:291]: Looks like you accidentally deleted the `@Test` annotation along with the `@SuppressWarnings`

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/GrpcServiceBuilder.java
Comment (jack-berg) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/GrpcServiceBuilder.java]: Just slightly more surface area to maintain. I don't feel strongly about this thought.

File: exporters/jaeger/src/main/java/io/opentelemetry/exporter/jaeger/JaegerGrpcSpanExporterBuilder.java
Comment (anuraaga) [inline exporters/jaeger/src/main/java/io/opentelemetry/exporter/jaeger/JaegerGrpcSpanExporterBuilder.java:104]:
```suggestion
      byte[] privateKeyPem, byte[] certificatePem) {
```

Throughout

File: exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/TlsUtilTest.java
Comment (anuraaga) [inline exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/TlsUtilTest.java:18]: It's sort of unclear what this is testing, I think we should remove this test and have actual usage of client TLS in our exporter test suite

https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractGrpcTelemetryExporterTest.java

We also won't need to check in any of these certificates and can use the selfsignedcertificate extension.

`tlsCustomizer` can be used to add trusted client certificates to the server

https://javadoc.io/doc/com.linecorp.armeria/armeria-javadoc/latest/com/linecorp/armeria/server/ServerBuilder.html

If it's unclear how to make the test work I can help by adding to this PR, let me know

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java
Comment (jkamon) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java:111]: since not providing both is not an error I will add condition to only call accept only in positive case when both are provided.  (case when both are not provided will be silently "ignored" as this is expected behaviour)

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java
Comment (jkamon) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java]: ok leaving early

File: exporters/jaeger/src/test/java/io/opentelemetry/exporter/jaeger/JaegerGrpcSpanExporterTest.java
Comment (jkamon) [reply exporters/jaeger/src/test/java/io/opentelemetry/exporter/jaeger/JaegerGrpcSpanExporterTest.java:292]: added back

File: exporters/jaeger/src/main/java/io/opentelemetry/exporter/jaeger/JaegerGrpcSpanExporterBuilder.java
Comment (jkamon) [reply exporters/jaeger/src/main/java/io/opentelemetry/exporter/jaeger/JaegerGrpcSpanExporterBuilder.java:104]: I can rename it here. Do you want me to rename it in all places where setClientTls is used?

File: exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/TlsUtilTest.java
Comment (jkamon) [reply exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/TlsUtilTest.java:18]: There are several parts I can see in this comment:

1. unclear test scope - It was an attempt to test also loading PEM file. 
2. I understand the point in not checking in the files at all and replace them with extension for certificates. I thought it might be useful to provide example to users how provided file should look like. It it should be with start/stop block, if it should be binary of base64 encoded. May be it wasn't the best idea. I can remove it / try to replace with tlsCustomizer
3. tlsCustomizer - will test less added code. 
4. I am not sure if got it completely. You want me to extend ServerBuilder with something like: 
`sb.tlsCustomizer(sslContextBuilder -> sslContextBuilder.keyManager(certificate.privateKey(), certificate.certificate()));`
? (should I create another key pair?)  If I am far from needed change in test then may be it would be better if you could help with test.

Comment (anuraaga) [reply exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/TlsUtilTest.java:18]: We need to add another selfsignedcertificate for what's used by the TLS-enabled client. 
`static final SelfSignedCertificateExtension clientCertificate = new SelfSignedCertificateExtension();`

We'll need to add that certificate as a trusted one for the server:
`sb.tlsCustomizer(ssl -> ssl.trustManager(clientCertificate.certificate())`
(BTW we don't want to change the server's TLS context as your snippet would do)

And then a test case for client TLS

```java
@Test
void clientTls() {
  exporterBuilder().setClientTls(Files.readAllBytes(clientCertificate.privateKey().toPath()), Files.readAllBytes(clientCertificate.certificate())

  // Similar to other tests
```

I think this will test what we want.

Do these hints help? If not I can add a commit with them.

Comment (jkamon) [reply exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/TlsUtilTest.java:18]: I think it does.
If it is still not what you expected then feel free to stop me and provide tests as you see fit.

File: exporters/jaeger/src/main/java/io/opentelemetry/exporter/jaeger/JaegerGrpcSpanExporterBuilder.java
Comment (anuraaga) [reply exporters/jaeger/src/main/java/io/opentelemetry/exporter/jaeger/JaegerGrpcSpanExporterBuilder.java:103]: Yeah

File: exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractGrpcTelemetryExporterTest.java
Comment (anuraaga) [inline exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractGrpcTelemetryExporterTest.java:262]: Ah does `certificate.certificate().getEncoded()` work same as next line?

File: exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/TlsUtilTest.java
Comment (anuraaga) [reply exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/TlsUtilTest.java:18]: Thanks! Let's go ahead and remove this test and related files, I think the server test looks swell.

**Review by anuraaga — COMMENTED** [review]
Thanks @jkamon - just small cleanup left. Though we'll probably want to hold off on merging until the spec PR is merged. OpenTelemetry is a frustratingly slow project, improving it will probably require a wide range of people making noise about it. So might be good to make noise at the spec repo or zoom meeting :)

File: exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/TlsUtilTest.java
Comment (jkamon) [reply exporters/otlp/common/src/test/java/io/opentelemetry/exporter/internal/TlsUtilTest.java]: done

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/TlsUtil.java
Comment (anuraaga) [inline exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/TlsUtil.java:146]: We can remove this function too now

File: exporters/jaeger/src/main/java/io/opentelemetry/exporter/jaeger/JaegerGrpcSpanExporterBuilder.java
Comment (jkamon) [reply exporters/jaeger/src/main/java/io/opentelemetry/exporter/jaeger/JaegerGrpcSpanExporterBuilder.java]: done

File: exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/TlsUtil.java
Comment (jkamon) [reply exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/TlsUtil.java]: done

**Comment by anuraaga** [comment]
LGTM - hopefully we can get the spec updated soon

**Comment by kubawach** [comment]
Triggered PR build.

**Review by anuraaga — APPROVED** [review]
Thanks for the patience @jkamon!

File: sdk-extensions/autoconfigure/README.md
Comment (jack-berg) [inline sdk-extensions/autoconfigure/README.md:70]: Can you document the log version of these as well for consistency? I know the spec doesn't mention the log version yet but we've already gone and added support / documentation for all the parameters so might as well be consistent.

**Review by jack-berg — APPROVED** [review]
Thanks for you patience!

File: sdk-extensions/autoconfigure/README.md
Comment (jkamon) [reply sdk-extensions/autoconfigure/README.md:71]: Done.
Can you please run the workflows? I would like to see if codecov will be OK this time or something else needs to be improved as well.

**Comment by jkamon** [comment]
@anuraaga @jack-berg @kubawach can one of you (or anyone else) please merge it before any conflicts will occur?

**Comment by jkamon** [comment]
@anuraaga Thanks!

**Pull request merged**
